x-pack/plugin/security/src/main/java/org/elasticsearch/xpack/security/authc/TokenService.java

```diff
@@ -1812,11 +1829,15 @@ static BytesReference createTokenDocument(
                     .field("invalidated", false)
                     .field("refreshed", false)
                     .startObject("client")
-                    .field("type", "unassociated_client")
-                    .field("user", originatingClientAuth.getUser().principal())
-                    .field("realm", originatingClientAuth.getAuthenticatedBy().getName());
-                if (originatingClientAuth.getAuthenticatedBy().getDomain() != null) {
-                    builder.field("realm_domain", originatingClientAuth.getAuthenticatedBy().getDomain());
+                    .field("type", "unassociated_client");
+                if (userToken.getVersion().onOrAfter(VERSION_CLIENT_AUTH_FOR_REFRESH)) {
+                    builder.field("authentication", originatingClientAuth.encode());
```

Comment:
>The authentication associated to userToken is downgraded with maybeRewriteForOlderVersion before it is passed into this method. I think we should apply the same logic tooriginatingClientAuth.

That is a good suggestion. Token documents get created in mixed-version clusters and it is important that every node that sees the token doc be able to make use of it.
I'll make it so as suggested, and also add a unit test somewhere for it.

> The version check logic means the new feature, "checking with authentication", is only used when min-node version is onOrAfter(8.2.0). Otherwise, the way of checking user, realm is applied even when all but a single node has not been upgraded to 8.2.0. I think this works and probably easier to explain. But I'd like to double confirm that is the intention.

That is indeed the intention; That's because downgrades are not supported, and a token created by any cluster node must be understood by any other cluster node.

Reply:
Pushed 51823f563a38df851884f2b605a1279 , but it is nasty to specifically test it.
Technically this should be test-covered by the ITs where a token created on the newer nodes in an upgrading cluster can be used on the older nodes (this is the only main code path that requires this change), and we do already have tests for that.